core/src/main/scala/kafka/metrics/clientmetrics/CmClientInstanceState.scala

```diff
@@ -0,0 +1,101 @@
+package kafka.metrics.clientmetrics
+
+import kafka.metrics.clientmetrics.ClientMetricsConfig.ClientMetrics.DEFAULT_PUSH_INTERVAL
+import kafka.metrics.clientmetrics.ClientMetricsConfig.SubscriptionGroup
+import org.apache.kafka.common.Uuid
+
+import java.nio.charset.StandardCharsets
+import java.util.Calendar
+import java.util.zip.CRC32
+import scala.collection.mutable.ListBuffer
+
+/**
+ * Client instance state that contains all the necessary information about the metric subscription for a client
+ */
+object CmClientInstanceState {
+
+  // Copy constructor
+  def apply(instance: CmClientInstanceState): CmClientInstanceState = {
+   create(instance.getId, instance.getClientInfo, instance.getSubscriptionGroups)
+  }
+
+  def apply(id: Uuid,
+            clientInfo: CmClientInformation,
+            cmGroups: java.util.Collection[SubscriptionGroup]): CmClientInstanceState = {
+    create(id, clientInfo, cmGroups)
+  }
+
+  private def create(id: Uuid,
+                     clientInfo: CmClientInformation,
+                     sgroups: java.util.Collection[SubscriptionGroup]): CmClientInstanceState = {
+
+    var targetMetrics = new ListBuffer[String]()
+    var pushInterval = DEFAULT_PUSH_INTERVAL
+    val targetGroups = new java.util.ArrayList[SubscriptionGroup]()
+    var allMetricsSubscribed = false
+
+    sgroups.forEach(v =>
```

Comment:
Ah, so this seems to be what I was thinking above about coalescing multiple subscriptions into a single subscription for the client.

Reply:
Yes this is where I compile all the metrics into a single list